src/shared_library.c

```diff
@@ -0,0 +1,158 @@
+  if (!lib->lib_pointer) {
+    lib->allocator.deallocate(lib->library_path, lib->allocator.state);
+    lib->library_path = NULL;
+    RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING("LoadLibrary error: %lu", GetLastError());
+    return RCUTILS_RET_ERROR;
+  }
+#endif  // _WIN32
+  return RCUTILS_RET_OK;
+}
+
+void *
+rcutils_get_symbol(const rcutils_shared_library_t * lib, const char * symbol_name)
+{
+  if (!lib || !lib->lib_pointer || (symbol_name == NULL)) {
+    RCUTILS_SET_ERROR_MSG("invalid inputs arguments");
+    return NULL;
+  }
+
+#ifndef _WIN32
+  void * lib_symbol = dlsym(lib->lib_pointer, symbol_name);
+  char * error = dlerror();
+  if (error != NULL) {
+    RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+      "Error getting the symbol '%s'. Error '%s'",
+      symbol_name, error);
+    return NULL;
+  }
+#else
+  void * lib_symbol = GetProcAddress(lib->lib_pointer, symbol_name);
+  if (lib_symbol == NULL) {
+    RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+      "Error getting the symbol '%s'. Error '%d'",
+      symbol_name, GetLastError());
+    return NULL;
+  }
+#endif  // _WIN32
+  if (!lib_symbol) {
+    RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+      "symbol '%s' doesnt' exit in library '%s'",
+      symbol_name, lib->library_path);
+    return NULL;
+  }
+  return lib_symbol;
+}
+
+bool
+rcutils_has_symbol(const rcutils_shared_library_t * lib, const char * symbol_name)
+{
+  if (!lib || !lib->lib_pointer || symbol_name == NULL) {
+    return false;
+  }
+
+#ifndef _WIN32
+  // the correct way to test for an error is to call dlerror() to clear any old error conditions,
+  // then call dlsym(), and then call dlerror() again, saving its return value into a variable,
+  // and check whether this saved value is not NULL.
+  dlerror(); /* Clear any existing error */
+  void * lib_symbol = dlsym(lib->lib_pointer, symbol_name);
+  return dlerror() == NULL && lib_symbol != 0;
+#else
+  void * lib_symbol = GetProcAddress(lib->lib_pointer, symbol_name);
+  return GetLastError() == 0 && lib_symbol != 0;
+#endif  // _WIN32
+}
+
+rcutils_ret_t
+rcutils_unload_shared_library(rcutils_shared_library_t * lib)
+{
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(lib, RCUTILS_RET_INVALID_ARGUMENT);
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(lib->lib_pointer, RCUTILS_RET_INVALID_ARGUMENT);
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(lib->library_path, RCUTILS_RET_INVALID_ARGUMENT);
+
+  rcutils_ret_t ret = RCUTILS_RET_OK;
+#ifndef _WIN32
+  // The function dlclose() returns 0 on success, and nonzero on error.
+  int error_code = dlclose(lib->lib_pointer);
+  if (error_code) {
+    RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING("dlclose error: %s", dlerror());
+    ret = RCUTILS_RET_ERROR;
+  }
+#else
+  // If the function succeeds, the return value is nonzero.
+  int error_code = FreeLibrary(lib->lib_pointer);
+  if (!error_code) {
+    RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING("FreeLibrary error: %lu", GetLastError());
+    ret = RCUTILS_RET_ERROR;
+  }
+#endif  // _WIN32
```

Comment:
Please deduplicate the parts which are the same.

Reply:
the only duplicate line between the `if` statements is
```
ret = RCUTILS_RET_ERROR;
```

Reply:
And the `if` and the closing parenthesis and the definition of the `error_code`...

Reply:
The `if` is not duplicated? The "declaration" of the `error_code` is the same (`int`), but the "definition" is not (as I understand the terms in this case), one calls `dlclose` and the other calls `FreeLibrary`.

I actually think this code is fine as-is and needs no more deduplication.